# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/ipc/DocAccessibleParent.cpp
# Commit: 95a2cae69fb9
# Full Hash: 95a2cae69fb9d3763a8a8aba8beba2a78fdd570b
# Author: James Teh <jteh@mozilla.com>
# Date: 2022-05-05 16:11:56
# Regressor Bug: 1766794
# File Overlap Count: 2
# Description:
#   Bug 1766794: Lazily cache HyperText offsets for RemoteAccessibles. r=eeejay
#   
#   We already had a similar cache in local HyperTextAccessible.
#   This improves performance significantly when walking the text attributes of a container with a large number of text leaf children, such as is encountered when using view source.
#   This patch unifies that cache across local and remote.
# ==============================================================================

diff -r 3f8a611b454e -r 95a2cae69fb9 accessible/ipc/DocAccessibleParent.cpp
--- a/accessible/ipc/DocAccessibleParent.cpp	Thu May 05 10:56:32 2022 +0000
+++ b/accessible/ipc/DocAccessibleParent.cpp	Thu May 05 11:14:12 2022 +0000
@@ -296,6 +296,10 @@
   if (StaticPrefs::accessibility_cache_enabled_AtStartup()) {
     if (aEventType == nsIAccessibleEvent::EVENT_REORDER ||
         aEventType == nsIAccessibleEvent::EVENT_INNER_REORDER) {
+      if (proxy->IsHyperText()) {
+        // Invalidate the HyperText offset cache.
+        proxy->InvalidateCachedHyperTextOffsets();
+      }
       for (RemoteAccessible* child = proxy->RemoteFirstChild(); child;
            child = child->RemoteNextSibling()) {
         child->InvalidateGroupInfo();
@@ -574,6 +578,13 @@
     }
 
     remote->ApplyCache(aUpdateType, entry.Fields());
+    if (aUpdateType == CacheUpdateType::Update && remote->IsTextLeaf()) {
+      // Invalidate the HyperText offset cache.
+      RemoteAccessible* parent = remote->RemoteParent();
+      if (parent && parent->IsHyperText()) {
+        parent->InvalidateCachedHyperTextOffsets();
+      }
+    }
   }
 
   if (nsCOMPtr<nsIObserverService> obsService =